File: examples/checks/queens.py
# Independent Python checks for the queens example.
from __future__ import annotations

import re

from .common import run_checks


def parse_columns(answer: str) -> list[int]:
    match = re.search(r"As column positions by row: \[([^\]]+)\]", answer)
    if not match:
        return []
    return [int(part.strip()) for part in match.group(1).split(",")]


def parse_total(answer: str) -> int | None:
    match = re.search(r"Total solutions for \d+-Queens: (\d+)", answer)
    return int(match.group(1)) if match else None


def count_n_queens(n: int) -> int:
    total = 0

    def search(row: int, cols: int, diag1: int, diag2: int) -> None:
        nonlocal total
        if row == n:
            total += 1
            return
        available = ((1 << n) - 1) & ~(cols | diag1 | diag2)
        while available:
            bit = available & -available
            available -= bit
            search(row + 1, cols | bit, (diag1 | bit) << 1, (diag2 | bit) >> 1)

    search(0, 0, 0, 0)
    return total


def run(ctx):
    data = ctx.load_input()
    n = int(data["N"])
    max_print = int(data["MaxPrint"])
    columns = parse_columns(ctx.answer)
    total = parse_total(ctx.answer)
    zero_based = [c - 1 for c in columns]
    diagonal_pairs = [
        (r1, r2)
        for r1 in range(len(zero_based))
        for r2 in range(r1 + 1, len(zero_based))
        if abs(zero_based[r1] - zero_based[r2]) == abs(r1 - r2)
    ]
    rendered_rows = [line for line in ctx.answer.splitlines() if set(line.strip().split()) <= {"Q", "."} and "Q" in line]
    independent_total = count_n_queens(n)

    checks = [
        ("the Python checker loaded the normalized 8-Queens input", n == 8 and max_print == 1),
        ("the printed solution gives one column for each row", len(columns) == n),
        ("all printed columns are within the board", all(1 <= c <= n for c in columns)),
        ("the printed solution uses each column exactly once", sorted(columns) == list(range(1, n + 1))),
        ("no pair of printed queens shares a diagonal", diagonal_pairs == []),
        ("the rendered board contains exactly eight rows and eight queens", len(rendered_rows) == n and sum(row.split().count("Q") for row in rendered_rows) == n),
        ("an independent Python bit-mask search counts 92 total solutions", independent_total == 92),
        ("the reported total matches the independent Python solution count", total == independent_total),
    ]
    return run_checks(checks)

File: examples/checks/queens.go
package main

import (
	"math"
	"sort"
	"strings"
)

func checkQueens(ctx *Context) []Check {
	d := ctx.M()
	n := integer(d["N"])
	maxp := integer(d["MaxPrint"])
	cols := []int{}
	if m := reFind(ctx.Answer, `As column positions by row: \[([^\]]+)\]`); m != nil {
		for _, p := range strings.Split(m[0], ",") {
			cols = append(cols, parseInt(strings.TrimSpace(p)))
		}
	}
	total := grabInt(ctx.Answer, `Total solutions for \d+-Queens: (\d+)`)
	zero := []int{}
	for _, c := range cols {
		zero = append(zero, c-1)
	}
	diagOK := true
	for r1 := 0; r1 < len(zero); r1++ {
		for r2 := r1 + 1; r2 < len(zero); r2++ {
			if int(math.Abs(float64(zero[r1]-zero[r2]))) == r2-r1 {
				diagOK = false
			}
		}
	}
	rows := 0
	queens := 0
	for _, line := range strings.Split(ctx.Answer, "\n") {
		fields := strings.Fields(line)
		if len(fields) > 0 {
			ok := true
			q := 0
			for _, f := range fields {
				if f != "Q" && f != "." {
					ok = false
				}
				if f == "Q" {
					q++
				}
			}
			if ok && q > 0 {
				rows++
				queens += q
			}
		}
	}
	indep := countQueens(n)
	sorted := append([]int(nil), cols...)
	sort.Ints(sorted)
	want := []int{}
	for i := 1; i <= n; i++ {
		want = append(want, i)
	}
	inRange := true
	for _, c := range cols {
		inRange = inRange && c >= 1 && c <= n
	}
	return []Check{{"the checker loaded the normalized 8-Queens input", n == 8 && maxp == 1}, {"the printed solution gives one column for each row", len(cols) == n}, {"all printed columns are within the board", inRange}, {"the printed solution uses each column exactly once", intsEq(sorted, want)}, {"no pair of printed queens shares a diagonal", diagOK}, {"the rendered board contains exactly eight rows and eight queens", rows == n && queens == n}, {"an independent bit-mask search counts 92 total solutions", indep == 92}, {"the reported total matches the independent solution count", total == indep}}
}
